src/mapblock.cpp

```diff
@@ -24,6 +24,77 @@
 #include "util/serialize.h"
 #include "util/basic_macros.h"
 
+// Like a std::unordered_map<content_t, content_t>, but faster.
+//
+// Unassigned entries are marked with 0xFFFF.
+//
+// The static memory requires about 65535 * 2 bytes RAM in order to be
+// sure we can handle all content ids.
+class IdIdMapping
+{
+	static_assert(sizeof(content_t) == 2, "content_t must be 16-bit");
+
+private:
+	std::unique_ptr<content_t[]> m_mapping;
+	std::vector<content_t> m_dirty;
+
+	static thread_local std::unique_ptr<content_t[]> tl_mapping;
+
+public:
+	IdIdMapping()
+	{
+		m_mapping = std::move(tl_mapping);
+		if (!m_mapping) {
+			m_mapping = std::make_unique<content_t[]>(CONTENT_MAX + 1);
+			memset(m_mapping.get(), 0xFF, (CONTENT_MAX + 1) * sizeof(content_t));
+		}
+	}
+
+	DISABLE_CLASS_COPY(IdIdMapping)
+
+	IdIdMapping(IdIdMapping &&other) noexcept :
+		m_mapping(std::move(other.m_mapping)), m_dirty(std::move(other.m_dirty))
+	{
+	}
+
+	IdIdMapping &operator=(IdIdMapping &&other) noexcept
+	{
+		if (&other == this)
+			return *this;
+		m_mapping = std::move(other.m_mapping);
+		m_dirty = std::move(other.m_dirty);
+		return *this;
+	}
+
+	content_t get(content_t k)
+	{
+		return m_mapping[k];
+	}
+
+	void set(content_t k, content_t v)
+	{
+		m_mapping[k] = v;
+		m_dirty.push_back(k);
+	}
+
+	void reset()
+	{
+		for (auto c : m_dirty)
+			m_mapping[c] = 0xFFFF;
+		m_dirty.clear();
+	}
+
+	~IdIdMapping()
+	{
+		if (!m_mapping)
+			return; // empty after move
+		reset();
+		tl_mapping = std::move(m_mapping);
+	}
+};
+
+thread_local std::unique_ptr<content_t[]> IdIdMapping::tl_mapping = nullptr;
```

Comment:
smells hacky. just make the `IdIdMapping` instance itself thread-local?
can even make it file-scope if wasting memory is a concern

Reply:
You mean a `thread_local IdIdMapping bla;`?
Would have to add another class then for the destructor, or explicitly give it back.

Reply:
I don't get what you mean.
just imagine this, no special handling required:
```cpp

static thread_local std::vector<T> cached_data;

void foo() {
    // do something with cached_data
}

void bar() {
    // do something with cached_data
}
```

Reply:
But we need to reset the cached data. If we just use it and forget to reset it in some return path (e.g. exception), it'll be in a dirty state.

Reply:
Sure, just call `reset` before. Nothing will happen if the array is not dirty.

Reply:
Like this you mean? https://github.com/luanti-org/luanti/pull/16000/commits/815f8552fba3a0102f1896cbe608b26502537fb5

Reply:
yes.

Reply:
I prefer the explicit borrow over the use of static variable, but whatever, both should be fine. :shrug: